src/source/parserutil.d

```diff
@@ -27,7 +27,17 @@ auto unexpectedTokenError(Lexer)(ref Lexer lexer, string expected) {
 		: format!"expected %s, not `%s`."(expected,
 		                                  token.toString(lexer.context));
 
+	const reachedEOF = lexer.index >= lexer.content.length - 1;
+
+	// End token indicates EOF between tokens
+	// reachedEOF checks for EOF within the token
```

Comment:
I suspect that reachedEOF is a suffiscient check here, as if EOF is not reached, no End token is emitted.